### DISCUSSION TITLE
Disable await state transfer by default to avoid timeouts when scaling the cluster

### DISCUSSION THREAD

**Discussion by pedroigor** [keycloak/keycloak#33636] [Ideas]
In a scenario where you have a huge number of (offline) sessions in caches and a cluster with auto-scaling, it might happen that multiple nodes fail to start and join the cluster if the initial state transfer that happens when bootstraping the distributable caches take too much time to complete.

A long start-up time in this case will cause issues when scaling automatically or force manual/additional steps to scale the cluster.

After talking with @pruivo, state transfer can be configured to:

* Disable state transfer
* Disable the timeout while transferring state in the background

I would like to know if it makes sense to at least disable the await state transfer timeout or even disable state transfer completely. As you know we have the database storing sessions and nodes can always load these sessions on-demand. 

The ultimate goal being to improve UX when deploying Keycloak by:

* Avoiding additional (manual) configuration to deploy the server
* Keep start-up time consistent and avoid errors when scaling a cluster
* Making it more aligned with production-ready deployments where auto-scaling is usually enabled

**Comment by ahus1** [comment]
The answer to your question depends on the Keycloak target version you are talking about. 

With Keycloak 26, the "huge number of (offline) sessions in caches" doesn't apply any more due to persistent sessions. By default it is set to 1 owner and 10_000 entries. Yes, we can disable state transfer here, and it will provide a small benefit, but not so much. 

With Keycloak 25, there were the options `spi-user-sessions-infinispan-offline-session-cache-entry-lifespan-override` and `spi-user-sessions-infinispan-offline-client-session-cache-entry-lifespan-override` where you could at least reduce the number of offline sessions kept in memory.

With Keycloak 24 and earlier, keeping all offline sessions in memory and Keycloak "soaking" up more and more sessions is a broken concept anyway, and the code needed some fixed that we added to later versions of Keycloak. As this is most affected by the problem, I wonder if you want to fix it for this version?

So what is the target version for the fix you're talking about?

**Reply by pedroigor** [reply]
As an upstream issue, I'm mainly talking about upstream code. If possible, improve this area in 26 too.

> With Keycloak 26, the "huge number of (offline) sessions in caches" doesn't apply any more due to persistent sessions. By default it is set to 1 owner and 10_000 entries. Yes, we can disable state transfer here, and it will provide a small benefit, but not so much.
>
>With Keycloak 25, there were the options spi-user-sessions-infinispan-offline-session-cache-entry-lifespan-override and spi-user-sessions-infinispan-offline-client-session-cache-entry-lifespan-override where you could at least reduce the number of offline sessions kept in memory.

The number of entries is configurable and the best value will depend on the use case. You might want to increase the number of entries in caches by thousands/millions while still keeping start up time consistent and non-blocking.

**Reply by ahus1** [reply]
The read operation on a database are usually quite fast and don't incur any IOPS. I'd first like to see a load test that extending the cache will give an improvement. At the moment, our Multi-Site setup runs without any Infinispan cache at all to avoid the overhead and complexities of invalidation.

**Reply by pruivo** [reply]
It is relevant for large clusters. I have no idea what is the average cluster size, but if it is in the double-digit (or triple-digit) sizes, a node join needs to receive state from all members to rebalance. Until the state transfer is finished, it won't accept any requests. 

At the end of the day, it is a decision if the user wants a cold cache (state transfer disabled) or a hot cache (state transfer enabled). Note that this choice is only available if persistent-user-session is enabled!

**Reply by pedroigor** [reply]
>It is relevant for large clusters. I have no idea what is the average cluster size, but if it is in the double-digit (or triple-digit) sizes, a node join needs to receive state from all members to rebalance.

Yeah, the use case that started this discussion is a 2 digit size cluster.

> Note that this choice is only available if persistent-user-session is enabled!
Wen using only offline sessions but not persistent-user-session too, right?

**Reply by pedroigor** [reply]
> The read operation on a database are usually quite fast and don't incur any IOPS. 
Yeah, that is why I think I good default would be to run state transfer in a non-blocking fashion. Or disable it completely. 

> I'd first like to see a load test that extending the cache will give an improvement. At the moment, our Multi-Site setup runs without any Infinispan cache at all to avoid the overhead and complexities of invalidation.

Btw, this is not really about runtime performance or multi-site. Running without a ISPN cache will actually solve the problem as you won't have state transfer blocking at start-up.

**Reply by pruivo** [reply]
> Wen using only offline sessions but not persistent-user-session too, right?

Yes, it is available for offline sessions since those are stored in the database all the time.

**Reply by ahus1** [reply]
Looking at this discussion, I would like to start with the expected the load and performance behavior the setup should handle (number of sessions, number of active sessions within a time interval, type of requests, number of users, etc.) and how/if Keycloak should scale up/down or how often it is restarted in rolling updates. 

Once those are known, we can test different architectures (persistent sessions with cache, persistent sessions without cache, etc.)

Having said all that, it might make sense to disable state transfer for the user session caches all the time. If we would agree on that, and also that this is not a breaking change, we could go ahead and change that for 26.1 
Still state transfers would occur for other caches like work cache or authenticationSessions cache.

**Reply by pedroigor** [reply]
This discussion started based on a use case when there was a quick peak load forcing the cluster to scale automatically. IIRC,the cluster scaled from 9 to 17 machines.

This is the error when a node fails to start and join the cluster:

```
1728213080400,"2024-10-06 08:11:20,199 ERROR [org.jboss.msc.service.fail] (ServerService Thread Pool -- 66) MSC000001: Failed to start service org.wildfly.clustering.infinispan.cache.keycloak.offlineClientSessions: org.jboss.msc.service.StartException in service org.wildfly.clustering.infinispan.cache.keycloak.offlineClientSessions: org.infinispan.commons.CacheException: Initial state transfer timed out for cache offlineClientSessions on ip-10-48-178-106.sa-east-1.compute.internal(site-id=sa-east-1a, rack-id=null, machine-id=43 bytes)
	at org.wildfly.clustering.service.FunctionalService.start(FunctionalService.java:66)
	at org.wildfly.clustering.service.AsyncServiceConfigurator$AsyncService.lambda$start$0(AsyncServiceConfigurator.java:117)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1348)
	at java.lang.Thread.run(Thread.java:826)
	at org.jboss.threads.JBossThread.run(JBossThread.java:513)"
1728213080400,"Caused by: org.infinispan.commons.CacheException: Initial state transfer timed out for cache offlineClientSessions on ip-10-48-178-106.sa-east-1.compute.internal(site-id=sa-east-1a, rack-id=null, machine-id=43 bytes)
	at org.infinispan.statetransfer.StateTransferManagerImpl.waitForInitialStateTransferToComplete(StateTransferManagerImpl.java:246)
	at org.infinispan.cache.impl.CacheImpl.start(CacheImpl.java:1013)
	at org.infinispan.cache.impl.AbstractDelegatingCache.start(AbstractDelegatingCache.java:512)
	at org.infinispan.manager.DefaultCacheManager.wireAndStartCache(DefaultCacheManager.java:697)
	at org.infinispan.manager.DefaultCacheManager.createCache(DefaultCacheManager.java:643)
	at org.infinispan.manager.DefaultCacheManager.internalGetCache(DefaultCacheManager.java:532)
	at org.infinispan.manager.DefaultCacheManager.getCache(DefaultCacheManager.java:510)
	at org.jboss.as.clustering.infinispan.DefaultCacheContainer.getCache(DefaultCacheContainer.java:85)
	at org.wildfly.clustering.infinispan.spi.service.CacheServiceConfigurator.get(CacheServiceConfigurator.java:77)
	at org.wildfly.clustering.infinispan.spi.service.CacheServiceConfigurator.get(CacheServiceConfigurator.java:55)
	at org.wildfly.clustering.service.FunctionalService.start(FunctionalService.java:63)
	... 7 more
```

The same error happened with the `offlineSessions` cache. The use case is mainly using offline sessions.

Even though the stack trace is based on Keycloak/Wildfly, the behavior should be the same for the current distribution.

**Reply by ahus1** [reply]
Did they scale a StatefulSet, or did they scale a Deployment? Just wondering if multiple nodes were started at the same time.

**Reply by pruivo** [reply]
Stretched clusters are always fun :) @pedroigor Ping me in the JIRA ticket if you need another pair of eyes.

**Comment by pedro-hos** [comment]
Hello all, I have a quick question. Looking into timeout configuration for RHSSO 7.6, we have a description for state transfer configuration.

~~~
"/subsystem=infinispan/cache-container=keycloak/distributed-cache=offlineSessions/component=state-transfer:read-resource-description()"

""The maximum amount of time (ms) to wait for state from neighboring caches, before throwing an exception and aborting startup. If timeout is 0, state transfer is performed asynchronously, and the cache will be immediately available.""
~~~

It seams that works like new Infinispan version features when using `awaitInitialTransfer=false`, the state transfer will be performed but in asynchronously way. Could you confirm that for us? Or by changing this, we will have the state transfer disabled?

**Reply by pruivo** [reply]
Yes, setting the `timeout=0` also sets `awaitInitialTransfer=false`.

source:
https://github.com/wildfly/wildfly/blob/dacb45d60249794978bd0c914f8c79189c2b7829/clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/StateTransferServiceConfigurator.java#L49-L57
